### PR TITLE
chore(actors): refactor allocation actor without actors

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change migrates the allocation actor to an 'allocation service' instead. The service mostly manages allocation structs that look similar to allocation actors, but have a clearer API---and the service's API should be much clearer. It also gets rid of a few stray APIs I was able to prove we don't need anymore (e.g., GetContainerResourcesState, which was not serving the purpose from the original PR that it was added for and was no longer needed).

## Test Plan
- [x] e2es
- [x] manual testing for restart

## Commentary (optional)
Urgh.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by erikwilson — APPROVED** [review]
Is a lot of code to go through but LGTM. Hopefully we can follow up on the hack and todo comments soon.

File: master/internal/task/allocation_service.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+		syslog.Warnf("dropped log for unknown allocation: %s", err)
+		return
+	}
+	ref.SendContainerLog(log)
+}
+
+// SetReady sets the ready bit and moves the allocation to the running state if it has not
+// progressed past it already.
+func (as *allocationService) SetReady(ctx context.Context, id model.AllocationID) error {
+	ref, err := as.waitForRestore(ctx, id)
```
Comment (NicholasBlaskey) [inline master/internal/task/allocation_service.go:103]: This isn't saying this is a problem just curious what we would need to do at this point to remove these ```waitForRestore```?

**Review by NicholasBlaskey — APPROVED** [review]
LGTM

Looks great couple minor questions and comments you can ignore.

File: tools/master.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+db:
```
[the remark this answers is not in the document]
Comment (stoksc) [reply tools/master.yaml:1]: oh, i just personally use them instead of `make devcluster` for testing fault tolerance

File: master/internal/task/allocation.go [hunk 3]
```diff
@@ -4,137 +4,165 @@ import (
...
+	Resources map[sproto.ResourcesID]sproto.ResourcesSummary
+	Ready     bool
+
+	Addresses  map[sproto.ResourcesID][]cproto.Address
+	Containers map[sproto.ResourcesID][]cproto.Container
+}
+
+// FirstContainer returns the first container in the allocation state.
+func (a AllocationState) FirstContainer() *cproto.Container {
+	for _, cs := range a.Containers {
```
Comment (maxrussell) [inline master/internal/task/allocation.go:63]: [Iteration order for maps isn't guaranteed](https://go.dev/blog/maps); is this meant to be a consistent order? Or is it meant to be the first container of an arbitrary map element? I guess I'm not really sure what these `ResourcesID`s are, so maybe there's really only ever going to be one.

Edit: wow, this just saw that this is pre-existing. Maybe a bug that should be fixed later?

File: master/internal/task/allocation.go [hunk 4]
```diff
@@ -4,137 +4,165 @@ import (
...
+type AllocationSignal string
 
 const (
-	killCooldown       = 15 * time.Second
-	okExitMessage      = "allocation exited successfully"
-	missingExitMessage = ""
+	// KillAllocation is the signal to kill an allocation; analogous to in SIGKILL.
+	KillAllocation AllocationSignal = "kill"
+	// TerminateAllocation is the signal to kill an allocation; analogous to in SIGTERM.
+	TerminateAllocation AllocationSignal = "terminate"
```
Comment (maxrussell) [inline master/internal/task/allocation.go:48]:
```suggestion
	// KillAllocation is the signal to kill an allocation; analogous to SIGKILL.
	KillAllocation AllocationSignal = "kill"
	// TerminateAllocation is the signal to kill an allocation; analogous to SIGTERM.
	TerminateAllocation AllocationSignal = "terminate"
```

File: master/internal/task/allocation.go [hunk 5]
```diff
@@ -171,224 +206,353 @@ func NewAllocation(
...
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.req.Restore && !a.restored
+}
+
+// waitForRestore waits until the allocation has been restored by the resource manager or a minute
+// has passed. If a minute passes, an error is returned. The allocation must exist otherwise this
+// will return a not found error.
+func (a *allocation) waitForRestore(ctx context.Context) error {
+	for i := 0; i < 60; i++ {
```
Comment (maxrussell) [inline master/internal/task/allocation.go:265]: Hmm, this waits for sixty iterations, but has no idea how long that's taken. If, for example, it blocks at `a.IsRestoring()` for awhile, it could wait much longer (which maybe is fine). Something like the following would be more accurate, but maybe that doesn't matter.
```go
waitStart := time.Now()
for ... {
    switch {
    //...
    }
    if time.Since(waitStart) > time.Minute {
        break
    } else {
        time.Sleep(time.Second)
    }
}
```

File: master/internal/task/allocation.go [hunk 6]
```diff
@@ -171,224 +206,353 @@ func NewAllocation(
...
-		Log: fmt.Sprintf("Scheduling %s (id: %s)", a.req.Name, ctx.Self().Parent().Address().Local()),
+		Log: fmt.Sprintf("Scheduling %s (id: %s)", a.req.Name, a.req.AllocationID),
 	})
-	return nil
+	return sub, nil
 }
 
-// Cleanup ensures an allocation is properly closed. It tries to do everything before failing and
+// Close ensures an allocation is properly closed. It tries to do everything before failing and
 // ensures we don't leave any resources running.
```
Comment (maxrussell) [inline master/internal/task/allocation.go:479]: Nit: in plain English, "closing" a door makes sense. Closing things like files or channels makes sense (we're done reading/writing). I don't know what closing an allocation means without reading the method's code. Maybe `CleanUp` was a better name? Or maybe we should beef up the doc comment.

**Review by maxrussell — APPROVED** [review]
Looks great! Nothing but nits.

File: master/Makefile [hunk 7]
```diff
@@ -2,7 +2,7 @@
 SHELL := bash
 GENERATED := packaging/LICENSE
 GENERATION_INPUTS = ../schemas/gen.py $(shell find ./pkg/schemas/ -name '*.go') $(shell find ../schemas/expconf -name '*.json')
-MOCK_GENERATION_INPUTS = ./internal/sproto/task.go ./internal/db/database.go ./internal/command/authz_iface.go
+MOCK_GENERATION_INPUTS = ./internal/sproto/task.go ./internal/db/database.go ./internal/command/authz_iface.go ./internal/rm/resource_manager_iface.go ./internal/task/allocation_service_iface.go
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/Makefile:5]: I agree. `moq` would fix this with inline gen comments.

[the remark this answers is not in the document]
Comment (stoksc) [reply master/Makefile:5]: not gonna do this now, but if you want, i'll stamp it.

File: master/internal/checkpoint_gc.go [hunk 8]
```diff
@@ -103,20 +101,24 @@ func (t *checkpointGCTask) Receive(ctx *actor.Context) error {
 			return fmt.Errorf("resolving resource pool: %w", err)
 		}
 
-		allocation := task.NewAllocation(t.logCtx, sproto.AllocateRequest{
+		err = task.DefaultService.StartAllocation(t.logCtx, sproto.AllocateRequest{
 			TaskID:            t.taskID,
 			JobID:             t.jobID,
+			RequestTime:       time.Now().UTC(),
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/checkpoint_gc.go:107]: It's one of those things that at this point seems dumb as a a request parameters but before didnt

File: master/internal/rm/rmevents/rmevents.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+type sequence struct {
+	mu sync.Mutex
+	i  int
+}
+
+func (s *sequence) next() int {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.i++
+	return s.i
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/rm/rmevents/rmevents.go:133]: sure. i'll just change it to a single atomic. didn't realize s.Add returned

File: master/internal/task/allocation.go (hunk 3 above)
Comment (stoksc) [reply master/internal/task/allocation.go:63]: Yeah, I think it is just to report telemetry. I'll rename it to `Single` in the orm-style for 'limit 1'.

File: master/Makefile (hunk 7 above)
[the remark this answers is not in the document]
Comment (maxrussell) [reply master/Makefile:5]: Since we have a ticket for spiking moq, I'll just leave it. Sorry I wasn't clear: this was just meant to be an observation of something weird, not a "hey, you should fix this."

File: master/internal/task/allocation.go (hunk 5 above)
Comment (stoksc) [reply master/internal/task/allocation.go:265]: copy pasta'd from old code, but yeah. going to just wrap the ctx with a timeout and use time.Ticker

File: master/internal/task/allocation.go (hunk 6 above)
Comment (stoksc) [reply master/internal/task/allocation.go:479]: renamed back, i agree. wanted to make it a proper close for other reasons but didnt get around to it.

File: master/internal/command/command.go [hunk 10]
```diff
@@ -348,10 +334,10 @@ func (c *command) Receive(ctx *actor.Context) error {
 		})
 	case *apiv1.KillNotebookRequest:
 		// TODO(Brad): Do the same thing to allocations that we are doing to RMs.
-		ctx.Tell(c.allocation, sproto.AllocationSignalWithReason{
-			AllocationSignal:    sproto.KillAllocation,
-			InformationalReason: "user requested kill",
-		})
+		err := task.DefaultService.Signal(c.allocationID, task.KillAllocation, "user requested kill")
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/command/command.go:342]: just keeping the current behavior for now. i figure it will be _really_ natural to do this as soon as i may command not an actor.

File: tools/master.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (stoksc) [reply tools/master.yaml:1]: i'll remove

File: master/internal/task/allocation_service.go (hunk 1 above)
Comment (stoksc) [reply master/internal/task/allocation_service.go:103]: uhh... i'm not sure. maybe once restore is instant. or it is gated globally and we can return 503 in middleware until restore is done, and make everything retry.

**Comment by stoksc** [comment]
(fyi this is brad borrowing rb's dev box, not rb)
